### PR TITLE
Add @nonlinear macro for modifying how expressions are parsed

### PR DESCRIPTION
**Pull request by odow**
x-ref https://github.com/jump-dev/JuMP.jl/issues/3498
x-ref https://github.com/jump-dev/JuMP.jl/issues/3729

 - [x] I'm open to bikeshedding the name (and whether we should even add this).

A core problem is that @mitchphillipson's models from #3729 contain `(x * (1 + y)) / c`, which creates an AffExpr `1 + y`, a QuadExpr `x * (1 + y)`, and then a new QuadExpr `(x * (1 + y)) / 2` (because `/(::Quad, ::Real)` is not in-place.

Since everything is about to be converted to `NonlinearExpr` anyway, it'd be nicer if we just kept things as `NonlinearExpr`.

cc @Downsite

### CODE REVIEW

**Comment by blegat** [comment]
> Since everything is about to be converted to `NonlinearExpr` anyway, it'd be nicer if we just kept things as `NonlinearExpr`.

In this case yes, but this performance issue could be hit in cases where a quadratic expression is the desired typed.
If the rewriting could would rewrite `(x * (1 + y)) / 2` into `MA.operate!!(/, x * (1 + y),  2)` then the performance issue would be fixed (provided the corresponding methods are defined in MA). I'd prefer fixing the performance issue, especially since it falls exactly within realms of issues MA was designed to fix.

**Comment by odow** [comment]
> but this performance issue could be hit in cases where a quadratic expression is the desired typed.

Exactly. Which is why this `@nl` is opt-in. We want to default to building affine and quadratic expressions. But sometimes people might want nonlinear ones.

It's not really about this specific performance issue. Even if `operate!(/, Float64, lhs, rhs)` was implemented, there would still be more temporaries allocated because creating `AffExpr` and QuadExpr` require allocating `OrderedDict` which are expensive.

**Comment by blegat** [comment]
I don't see any objection then, I can see this being useful. About the naming, once we drop the legacy nonlinear interface, what would be left of the `nl` shortcut ? If not much then maybe the `nl` shortcut is not so standard anymore in JuMP and we can use `@nonlinear`

**Comment by odow** [comment]
Yeah I have no strong opinions on the name. I don't like `@nl(...)`. `@nonlinear` could work.

**Comment by odow** [comment]
Bikeshed

 * `@_nl` ?
 * `@nl_esc`
 * `@scary_nl`
 * `@nl_boo` (@matbesancon)
 * `@nlexpr(...)`
 * `@nl_expr(...)`
 * `@raw_nl_expr(...)`
 * `@raw_nonlinear_expr(...)`
 * `@raw_nl`
 * ...

**Comment by odow** [comment]
This could use a review. I've updated to `@force_nonlinear` as discussed.

**Review by pulsipher — APPROVED** [review]
Looks good to me.

**Review by matbesancon — APPROVED** [review]

File: docs/src/manual/nonlinear.md [hunk 1]
```diff
@@ -355,6 +355,34 @@ julia> expr.args
...
+
+julia> @variable(model, x);
+
+julia> f = (x - 0.1)^2
+x² - 0.2 x + 0.010000000000000002
+
+julia> typeof(f)
+QuadExpr (alias for GenericQuadExpr{Float64, GenericVariableRef{Float64}})
+```
+To over-ride this behavior, use the [`@force_nonlinear`](@ref) macro:
```
Comment (matbesancon) [inline docs/src/manual/nonlinear.md:373]: nitpick: spelling override without hyphen?

File: src/macros/@force_nonlinear.jl [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                return Expr(:call, _op_div, x.args[2:end]...)
+            elseif x.args[1] == :^
+                return Expr(:call, _op_pow, x.args[2:end]...)
+            end
+        end
+        return x
+    end
+    return quote
+        r = $(esc(ret))
+        if !(r isa $GenericNonlinearExpr)
```
Comment (blegat) [inline src/macros/@force_nonlinear.jl:113]: I'd rather call a `_check_nonlinear` than complicate the generated code ? Less worry about hygiene issue of `r`, etc...

Comment (odow) [reply src/macros/@force_nonlinear.jl]: Done

**Review by blegat — APPROVED** [review]

**Comment by MAiNGO-github** [comment]
@odow Is it expected that @force_nonlinear does not work when the quadratic expression comes from a user-defined function?

`@force_nonlinear(p.funcs.g(x, w[j, :])): expression did not produce a GenericNonlinearExpr. Got a QuadExpr: w[1,1]*w[1,2] - x[1]*w[1,2] + w[1,1] + 0 x[1]")`

**Comment by Downsite** [comment]
@odow Is it expected that @force_nonlinear does not work when the quadratic expression comes from a user-defined function?

`@force_nonlinear(p.funcs.g(x, w[j, :])): expression did not produce a GenericNonlinearExpr. Got a QuadExpr: w[1,1]*w[1,2] - x[1]*w[1,2] + w[1,1] + 0 x[1]")`

**Comment by odow** [comment]
Yes, because the macro cannot rewrite the expressions inside `p.funcs.g`. 

We considered automatically converting the result, but decided against it.

You should modify `p.funcs.g` to create the appropriate expression type.

**Pull request merged**